### ISSUE TITLE
limit on SPARQL results?

### ISSUE THREAD

**Issue by michalovadek** [EUCouncil/CouncilVotesOnActsDatasetSample#2]
Hi, I wasn't sure about the best way to reach out, so I am trying here. I noticed that I get 100 000 rows when I try to query all votes on all acts using SPARQL. I suppose this is the maximum imposed on the server side. Is there a way to retrieve all the observations at the act-countryvote level in one query? The data in this structure are interesting for researchers, because you can enrich it at both the act and country levels.

And perhaps just to check, is the query below appropriate for what I am trying to do in the first place? (adapted from your examples)

`SELECT ?observation ?act ?country ?vote ?area ?title ?date ?rule
    where {
    ?observation
    <http://data.consilium.europa.eu/data/public_voting/qb/dimensionproperty/act>
    ?act
    .
    ?observation
    <http://data.consilium.europa.eu/data/public_voting/qb/measureproperty/vote>
    ?vote
    .
    ?observation
    <http://data.consilium.europa.eu/data/public_voting/qb/dimensionproperty/country>
    ?country
    .
    ?act
    skos:definition
    ?title
    .
    ?observation
    <http://data.consilium.europa.eu/data/public_voting/qb/dimensionproperty/actdate>
    ?date
    .
    ?observation
    <http://data.consilium.europa.eu/data/public_voting/qb/dimensionproperty/policyarea>
    ?area
    .
    ?observation
    <http://data.consilium.europa.eu/data/public_voting/qb/dimensionproperty/votingrule>
    ?rule
    .
    }`

Thanks in advance.

**Comment by simonech** [comment]
Hi,
I'm sorry for the late answer.
This repository is pretty old and has never been updated after the Hackathon.
I've to say I'm a bit rusty on the SPARQL query, so cannot really help at the moment.
I'll ping my colleague that is following up with this project at the moment.

@singuch: can you please see if you can help with this questions?
Thx

**Comment by singuch** [comment]
Hi,
As my colleague @simonech mentioned, this is old. The data set used in this repository was formatted as a cube which is easier to use with statistical tools and by the way the data set is not up to date. In the meantime we have published a new version of the data set which is formatted differently and updated regularly.

As you are interested to get all the information in one query, I'll give you a query that loads most of the data from the new data set. Each result line is a VotingProcedure (Act) and countries votes are grouped in columns (InFavour/Against/Abstained/NotParticipating).

This new data set is linked to another data set called PublicRegister that contains metadata about documents published by the Council. The "linked" part is used to retrieve the title. Sometimes it is difficult to identify the document or the document is not available thus the title is sometimes empty.

Here is the query: I added it as attached file to keep the formatting, that way it is easier to read.

[VR-sparql.txt](https://github.com/EUCouncil/CouncilVotesOnActsDatasetSample/files/5061036/VR-sparql.txt)

Kind regards,
Christophe

**Comment by michalovadek** [comment]
Thanks a lot Christophe for this input. I should have specified that I am trying to use this on the https://data.consilium.europa.eu/sparql endpoint rather than the cube included as part of this repository (I assumed the projects are closely linked).

I tried running the query but I get a syntax error. A SPARQL query validator identifies the following issue: "Encountered " "group_concat" "group_concat "" at line 9, column 258." I tried debugging myself with parentheses and braces but no luck so far.

**Comment by singuch** [comment]
You're welcome Michal.
How do you query the endpoint? Do you use a client or code on your side that connects to the endpoint or do you use the GUI of the endpoint (by opening https://data.consilium.europa.eu/sparql directly in the browser)?

If you use code or a client app:
- make sure it is using POST requests to call the endpoint otherwise the query can be truncated if you use GET requests;
- pay attention that the query contains single quotes characters that will have to be escaped if you store the query in your application code.

If you use the GUI of the endpoint, then it is strange, I downloaded the VR-sparql.txt file and executed the query and it worked fine.

Christophe

**Comment by michalovadek** [comment]
Thanks for your help, Christophe. You are absolutely right that the query works through the endpoint GUI. Which makes it all the more puzzling why I can't get it to work through my client. I am using a cURL-based solution in a statistical software called R. So far I had no issues implementing SPARQL requests with the same set-up using the EU Publications Office endpoint, so I was hoping for similar success with the Council endpoint.

I have so far tried both POST and GET. I tried escaping the 's. I tried both via URL encoding and including the query in the body (following https://www.w3.org/TR/sparql11-protocol/). I tried including various headers (e.g. 'Content-Type' = 'application/sparql-query'). But no combination of these parameters has so far been successful for me.

When I do POST with query in the body:
-> POST /sparql HTTP/1.1
-> Host: data.consilium.europa.eu
-> User-Agent: libcurl/7.64.1 r-curl/4.3 httr/1.4.1
-> Accept-Encoding: deflate, gzip
-> Accept: application/json, text/xml, application/xml, */*
-> Content-Length: 4416
... *your SPARQL query*

I get back simply the website of the GUI endpoint:
<- HTTP/1.1 200 OK
<- Content-Type: text/html; charset=UTF-8

When I add a Content-Type header:
-> POST /sparql HTTP/1.1
-> Host: data.consilium.europa.eu
-> User-Agent: libcurl/7.64.1 r-curl/4.3 httr/1.4.1
-> Accept-Encoding: deflate, gzip
-> Accept: application/json, text/xml, application/xml, */*
-> Content-Type: application/sparql-query
-> Content-Length: 4416
... *your SPARQL query*

I get back a 409:
<- HTTP/1.1 409 Invalid path
<- Content-Type: text/html; charset=UTF-8

I suppose there is something missing from my requests but I can't figure out what it is. Would you have any clue as to what might be going on here? Is there perhaps a need to specify the default graph URI? (I don't see it filled in the GUI)

**Comment by singuch** [comment]
Hi Michal,
Specifying the default graph URI is not needed, the query contains the following FROM clauses that indicates which datasets are targeted:
- FROM <http://data.consilium.europa.eu/id/dataset/VotingResults>
- FROM <http://data.consilium.europa.eu/id/dataset/PublicRegister>

I never did anything related to R and I'm using cUrl only for very basic things.
But here is some more information that may help you.
I'm querying data from the endpoint from an Angular application. The application is loading the data in JSON format.

To load the information, I'm adding to my request the following Content-Type header: '`application/x-www-form-urlencoded;charset=utf-8;`'

I'm also adding the following Http parameters:
- `default-graph-uri` with an empty value ('')
- `query` that contains the sparql query, (it must be url encoded)
- `format` with value `application/sparql-results+json`
- `timeout` with value `0`
- `debug`, with value '`on`'
those parameters are added to the body as we use POST requests.

In case you're interested to have a look at the application and inspect the requests (using some debug tools) that are done to the endpoint, you can have a look there: https://www.consilium.europa.eu/en/general-secretariat/corporate-policies/transparency/open-data/voting-results/

I found the following blog post that may help you too when using sparql with cUrl: http://www.snee.com/bobdc.blog/2019/02/curling-sparql.html

Keep me posted, maybe will I be able to help more if you get other error messages.

**Comment by michalovadek** [comment]
Thanks for all your help, Christophe. It seems I solved the problem. There seems to have been an issue in the way the arguments were specified in my R code. I went back first to pure cURL in the command prompt where I identified this simple code to do the trick:

`> curl --data-urlencode query@VR-sparql.txt https://data.consilium.europa.eu/sparql`

and in R:

`httr::POST(url = endpoint,
           body = list(query = query)
)`

Somehow I missed this very straightforward solution when I was trying different parameters and URL encodings.

I will now add this query and code to my [R package](https://github.com/michalovadek/eurlex) where I collate functions for downloading EU data. I will credit you with the solution. Thanks again!
